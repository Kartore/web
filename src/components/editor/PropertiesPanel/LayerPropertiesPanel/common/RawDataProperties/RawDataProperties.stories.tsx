import type { Meta, StoryObj } from '@storybook/react';

import { RawDataProperties } from '.';

const meta = {
  component: RawDataProperties,
} satisfies Meta<typeof RawDataProperties>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    layer: {
      id: 'circle-layer-test',
      type: 'circle',
      source: 'source-test',
      'source-layer': 'source-layer-test',
      minzoom: 0,
      maxzoom: 24,
      paint: {
        'circle-color': '#ffffff',
      },
    },
  },
};
