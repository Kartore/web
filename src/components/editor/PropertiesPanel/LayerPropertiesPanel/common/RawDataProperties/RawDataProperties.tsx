import type { ComponentProps, FC } from 'react';

import type { LayerSpecification } from '@maplibre/maplibre-gl-style-spec';
import { Editor } from '@monaco-editor/react';
import { shikiToMonaco } from '@shikijs/monaco';
import { createHighlighter } from 'shiki/bundle/web';

import type { onChangeType } from '~/components/editor/PropertiesPanel/LayerPropertiesPanel/utils/LayerUtil/LayerUtil.ts';
import { cn } from '~/utils/tailwindUtil';

export type RawDataPropertiesProps = {
  layer: LayerSpecification;
  onChange?: onChangeType;
} & Omit<ComponentProps<'div'>, 'onChange'>;

export const RawDataProperties: FC<RawDataPropertiesProps> = ({
  layer,
  className,
  children,
  onChange,
  ...props
}) => {
  return (
    <div {...props} className={cn('flex flex-col gap-2 px-4', className)}>
      <h3 className={'font-montserrat text-sm font-semibold'}>Raw Data Editor (Advanced)</h3>

      <Editor
        className={cn('min-h-40', className)}
        value={JSON.stringify(layer, undefined, 2)}
        defaultLanguage={'json'}
        onChange={(value) => {
          if (onChange && value) onChange(layer, 'all', undefined, JSON.parse(value));
        }}
        options={{
          minimap: { enabled: false },
        }}
        onMount={(_, monaco) => {
          (async () => {
            const highlighter = await createHighlighter({
              themes: ['light_plus'],
              langs: ['json'],
            });
            monaco.languages.register({ id: 'json' });
            shikiToMonaco(highlighter, monaco);
          })();
        }}
      />
      {children}
    </div>
  );
};

RawDataProperties.displayName = 'RawDataProperties';
