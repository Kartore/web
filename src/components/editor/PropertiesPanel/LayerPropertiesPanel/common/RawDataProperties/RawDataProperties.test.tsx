import { describe } from 'vitest';
import { render } from '@testing-library/react';
import { composeStories } from '@storybook/react';

import * as Stories from './RawDataProperties.stories';

const { Default } = composeStories(Stories);

describe('Component: RawDataProperties', () => {
  describe('Snapshot', () => {
    it('Default', () => {
      const { asFragment } = render(<Default />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
