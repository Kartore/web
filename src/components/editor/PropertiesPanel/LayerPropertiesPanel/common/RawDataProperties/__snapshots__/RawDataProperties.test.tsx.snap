// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component: RawDataProperties > Snapshot > Default 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-2 px-4"
  >
    <h3
      class="font-montserrat text-sm font-semibold"
    >
      Raw Data Editor (Advanced)
    </h3>
    <textarea
      class="min-h-40 rounded text-sm"
      readonly=""
    >
      {
  "id": "circle-layer-test",
  "type": "circle",
  "source": "source-test",
  "source-layer": "source-layer-test",
  "minzoom": 0,
  "maxzoom": 24,
  "paint": {
    "circle-color": "#ffffff"
  }
}
    </textarea>
  </div>
</DocumentFragment>
`;
